ext/aes256gcm_decrypt/aes256gcm_decrypt.c

```diff
@@ -43,55 +50,60 @@ VALUE method_aes256gcm_decrypt_decrypt(VALUE self, VALUE rb_ciphertext, VALUE rb
 
   /* Initialise decryption operation */
   if (!EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL)) {
-    goto cleanup1;
+    goto cleanup3;
   }
 
   /* Set initialisation vector (IV) length */
   if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, iv_len, NULL)) {
-    goto cleanup1;
+    goto cleanup3;
   }
 
   /* Initialise key and IV */
   if (!EVP_DecryptInit_ex(ctx, NULL, NULL, key, iv)) {
-    goto cleanup1;
+    goto cleanup3;
   }
 
   /* Provide empty string as additional authenticated data (AAD) */
-  if (!EVP_DecryptUpdate(ctx, NULL, &len, "", 0)) {
-    goto cleanup1;
+  if (!EVP_DecryptUpdate(ctx, NULL, &len, (unsigned char *)"", 0)) {
+    goto cleanup3;
   }
 
   /* Provide message to be decrypted */
   if (!EVP_DecryptUpdate(ctx, plaintext, &len, ciphertext, ciphertext_len)) {
-    goto cleanup1;
+    goto cleanup3;
   }
   plaintext_len = len;
 
   /* Set expected tag */
   if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, tag_len, tag)) {
-    goto cleanup1;
+    goto cleanup3;
   }
 
   /* Finalise decryption */
   if (EVP_DecryptFinal_ex(ctx, plaintext + len, &len) > 0) {
```

Comment:
I'm not fond of non-explicit pointer arithmetic. I would prefer something like
```C
if (EVP_DecryptFinal_ex(ctx, &plaintext[len], &len) > 0) {
```

Reply:
Diving into this led me to http://caxapa.ru/thumbs/468328/misra-c-2004.pdf , i.e. "Guidelines for the use of the C language in critical systems" (via [this blog post](https://critical.eschertech.com/2010/03/03/how-unsafe-is-pointer-arithmetic/)). Rule 17.4 on PDF page 72, marked as "required", states:

> Array indexing shall be the only allowed form of pointer arithmetic

As mentioned in the blog post, 17.4 appears to make rules 17.1-17.3 redundant :slightly_smiling_face: 

It is probably a good idea to adhere to that rule.